Expose https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom

youennf opened the pull request:
#### 1afb53f1cc52623f60845f1c57acc703bca8bfa8
<pre>
Expose <a href="https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom">https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=253646">https://bugs.webkit.org/show_bug.cgi?id=253646</a>
rdar://problem/106498125

Reviewed by Eric Carlson.

Add support for <a href="https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom.">https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom.</a>
Expose zoom as a supported constraint.
In AVVideoCaptureSource, we currently restrict zoom to 4 at most to limit potential privacy issue, since controlling zoom is supposed to require higher camera priviledges.
This requires some changes to RealtimeVideoCaptureSource since we set size, frame rate and zoom at the same time.

Update mock camera to support zooming and add new test to cover this.
When mock camera is using zoom, we set the background to red.

* LayoutTests/fast/mediastream/MediaDevices-getSupportedConstraints-expected.txt:
* LayoutTests/fast/mediastream/MediaDevices-getSupportedConstraints.html:
* LayoutTests/fast/mediastream/mediastreamtrack-video-zoom-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-zoom.html: Added.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
(WebCore::MediaDevices::getSupportedConstraints):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
(WebCore::MediaStreamTrack::getCapabilities const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::zoom const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::supportsSizeFrameRateAndZoom):
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::selectSettings):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::supportsConstraints):
(WebCore::RealtimeMediaSource::applyConstraints):
(WebCore::RealtimeMediaSource::setZoom):
(WebCore::RealtimeMediaSource::supportsSizeAndFrameRate): Deleted.
(WebCore::RealtimeMediaSource::setSizeAndFrameRate): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsZoom const):
(WebCore::RealtimeMediaSourceCapabilities::zoom const):
(WebCore::RealtimeMediaSourceCapabilities::setZoom):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::RealtimeMediaSourceCenter):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::RealtimeMediaSourceSettings::difference const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsZoom const):
(WebCore::RealtimeMediaSourceSettings::zoom const):
(WebCore::RealtimeMediaSourceSettings::setZoom):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsZoom const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsZoom):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::updateCapabilities):
(WebCore::RealtimeVideoCaptureSource::supportsSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::presetSupportsFrameRate):
(WebCore::RealtimeVideoCaptureSource::presetSupportsZoom):
(WebCore::zoomFromPreset):
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::clientUpdatedSizeFrameRateAndZoom):
(WebCore::SizeFrameRateAndZoom::toJSONObject const):
(WebCore::SizeFrameRateAndZoom::toJSONString const):
(WebCore::RealtimeVideoCaptureSource::supportsSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoCaptureSource::setSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoCaptureSource::clientUpdatedSizeAndFrameRate): Deleted.
(WebCore::SizeAndFrameRate::toJSONObject const): Deleted.
(WebCore::SizeAndFrameRate::toJSONString const): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
(WTF::LogArgument&lt;WebCore::SizeFrameRateAndZoom&gt;::toString):
(WTF::LogArgument&lt;WebCore::SizeAndFrameRate&gt;::toString): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoSource.cpp:
(WebCore::RealtimeVideoSource::supportsSizeFrameRateAndZoom):
(WebCore::RealtimeVideoSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeVideoSource::settingsDidChange):
(WebCore::RealtimeVideoSource::supportsSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoSource::setSizeAndFrameRate): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoSource.h:
* Source/WebCore/platform/mediastream/VideoPreset.h:
(WebCore::VideoPreset::create):
(WebCore::VideoPreset::VideoPreset):
(WebCore::VideoPreset::log const):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::computeMaxZoom):
(WebCore::AVVideoPreset::AVVideoPreset):
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::setFrameRateAndZoomWithPreset):
(WebCore::AVVideoCaptureSource::setSessionSizeFrameRateAndZoom):
(WebCore::AVVideoCaptureSource::setupCaptureSession):
(WebCore::AVVideoCaptureSource::setFrameRateWithPreset): Deleted.
(WebCore::AVVideoCaptureSource::setSessionSizeAndFrameRate): Deleted.
* Source/WebCore/platform/mock/MockMediaDevice.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::supportsSizeFrameRateAndZoom):
(WebCore::MockRealtimeVideoSource::setSizeFrameRateAndZoom):
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
(WebCore::MockRealtimeVideoSource::setFrameRateAndZoomWithPreset):
(WebCore::MockRealtimeVideoSource::generateFrame):
(WebCore::MockRealtimeVideoSource::supportsSizeAndFrameRate): Deleted.
(WebCore::MockRealtimeVideoSource::setSizeAndFrameRate): Deleted.
(WebCore::MockRealtimeVideoSource::setFrameRateWithPreset): Deleted.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/261679@main">https://commits.webkit.org/261679@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c89d86b067c7f66544ee6ccfe014933ef510c81c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112543 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21693 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1198 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4317 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121099 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116609 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23031 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12854 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5455 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118310 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100310 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105604 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/855 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46113 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14031 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/892 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14713 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20052 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52897 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8138 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16549 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -356,7 +356,7 @@ bool RealtimeMediaSource::supportsSizeAndFrameRate(std::optional<IntConstraint>
         if (std::isinf(constraintDistance)) {
             auto range = capabilities.width();
-            WTFLogAlways("RealtimeMediaSource::supportsSizeAndFrameRate failed width constraint, capabilities are [%d, %d]", range.rangeMin().asInt, range.rangeMax().asInt);
+            WTFLogAlways("RealtimeMediaSource::supportsSizeFrameRateAndZoom failed width constraint, capabilities are [%d, %d]", range.rangeMin().asInt, range.rangeMax().asInt);
```
We should really use `ERROR_LOG` instead of `WTFLogAlways` to log errors in this class so they show up in logs filtered with the standard WebRTC predicate.

youennf replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=253903 to do it as a follow-up

eric-carlson reviewed (approved):
r=me

youennf committed:
Expose https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom
https://bugs.webkit.org/show_bug.cgi?id=253646
rdar://problem/106498125

Reviewed by Eric Carlson.

Add support for https://w3c.github.io/mediacapture-image/#dom-mediatrackcapabilities-zoom.
Expose zoom as a supported constraint.
In AVVideoCaptureSource, we currently restrict zoom to 4 at most to limit potential privacy issue, since controlling zoom is supposed to require higher camera priviledges.
This requires some changes to RealtimeVideoCaptureSource since we set size, frame rate and zoom at the same time.

Update mock camera to support zooming and add new test to cover this.
When mock camera is using zoom, we set the background to red.

* LayoutTests/fast/mediastream/MediaDevices-getSupportedConstraints-expected.txt:
* LayoutTests/fast/mediastream/MediaDevices-getSupportedConstraints.html:
* LayoutTests/fast/mediastream/mediastreamtrack-video-zoom-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-zoom.html: Added.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
(WebCore::MediaDevices::getSupportedConstraints):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
(WebCore::MediaStreamTrack::getCapabilities const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::zoom const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::supportsSizeFrameRateAndZoom):
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::selectSettings):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::supportsConstraints):
(WebCore::RealtimeMediaSource::applyConstraints):
(WebCore::RealtimeMediaSource::setZoom):
(WebCore::RealtimeMediaSource::supportsSizeAndFrameRate): Deleted.
(WebCore::RealtimeMediaSource::setSizeAndFrameRate): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsZoom const):
(WebCore::RealtimeMediaSourceCapabilities::zoom const):
(WebCore::RealtimeMediaSourceCapabilities::setZoom):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::RealtimeMediaSourceCenter):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::RealtimeMediaSourceSettings::difference const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsZoom const):
(WebCore::RealtimeMediaSourceSettings::zoom const):
(WebCore::RealtimeMediaSourceSettings::setZoom):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsZoom const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsZoom):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::updateCapabilities):
(WebCore::RealtimeVideoCaptureSource::supportsSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::presetSupportsFrameRate):
(WebCore::RealtimeVideoCaptureSource::presetSupportsZoom):
(WebCore::zoomFromPreset):
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::clientUpdatedSizeFrameRateAndZoom):
(WebCore::SizeFrameRateAndZoom::toJSONObject const):
(WebCore::SizeFrameRateAndZoom::toJSONString const):
(WebCore::RealtimeVideoCaptureSource::supportsSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoCaptureSource::setSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoCaptureSource::clientUpdatedSizeAndFrameRate): Deleted.
(WebCore::SizeAndFrameRate::toJSONObject const): Deleted.
(WebCore::SizeAndFrameRate::toJSONString const): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
(WTF::LogArgument<WebCore::SizeFrameRateAndZoom>::toString):
(WTF::LogArgument<WebCore::SizeAndFrameRate>::toString): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoSource.cpp:
(WebCore::RealtimeVideoSource::supportsSizeFrameRateAndZoom):
(WebCore::RealtimeVideoSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeVideoSource::settingsDidChange):
(WebCore::RealtimeVideoSource::supportsSizeAndFrameRate): Deleted.
(WebCore::RealtimeVideoSource::setSizeAndFrameRate): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoSource.h:
* Source/WebCore/platform/mediastream/VideoPreset.h:
(WebCore::VideoPreset::create):
(WebCore::VideoPreset::VideoPreset):
(WebCore::VideoPreset::log const):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::computeMaxZoom):
(WebCore::AVVideoPreset::AVVideoPreset):
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::setFrameRateAndZoomWithPreset):
(WebCore::AVVideoCaptureSource::setSessionSizeFrameRateAndZoom):
(WebCore::AVVideoCaptureSource::setupCaptureSession):
(WebCore::AVVideoCaptureSource::setFrameRateWithPreset): Deleted.
(WebCore::AVVideoCaptureSource::setSessionSizeAndFrameRate): Deleted.
* Source/WebCore/platform/mock/MockMediaDevice.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::supportsSizeFrameRateAndZoom):
(WebCore::MockRealtimeVideoSource::setSizeFrameRateAndZoom):
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
(WebCore::MockRealtimeVideoSource::setFrameRateAndZoomWithPreset):
(WebCore::MockRealtimeVideoSource::generateFrame):
(WebCore::MockRealtimeVideoSource::supportsSizeAndFrameRate): Deleted.
(WebCore::MockRealtimeVideoSource::setSizeAndFrameRate): Deleted.
(WebCore::MockRealtimeVideoSource::setFrameRateWithPreset): Deleted.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/261679@main
